File: 034__Find_First_and_Last_Position_of_Element_in_Sorted_Array.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        def findmax(l, r):
            while l < r:
                mid = (l + r) // 2
                if nums[mid] <= target:
                    l = mid + 1
                else:
                    r = mid
            return l - 1

        def findmin(l, r):
            while l < r:
                mid = (l + r) // 2
                if nums[mid] < target:
                    l = mid + 1
                else:
                    r = mid
            return l

        if len(nums) == 0:
            return [-1, -1]

        index_l = findmin(0, len(nums))
        if index_l == len(nums) or nums[index_l] != target:
            return [-1, -1]

        return [index_l, findmax(0, len(nums))]

File: Java/034__Find_First_and_Last_Position_of_Element_in_Sorted_Array.java
class Solution {

    public int[] searchRange(int[] nums, int target) {
    		int start = firstGreaterEqual(nums, target);
    		if (start == nums.length || nums[start] != target) {
      			return new int[]{-1, -1};
    		}
    		return new int[]{start, firstGreaterEqual(nums, target + 1) - 1};
	}

	//find the first number that is greater than or equal to target.
	//could return A.length if target is greater than A[A.length-1].
	//actually this is the same as lower_bound in C++ STL.
	private static int firstGreaterEqual(int[] nums, int target) {
		int low = 0, high = nums.length;
		while (low < high) {
			int mid = (low + high) / 2;
			//low <= mid < high
			if (nums[mid] < target) {
				low = mid + 1;
			} else {
				//should not be mid-1 when A[mid]==target.
				//could be mid even if A[mid]>target because mid<high.
				high = mid;
			}
		}
		return low;
    }
}
